pybind11/setup_helpers.py

```diff
@@ -0,0 +1,243 @@
+# -*- coding: utf-8 -*-
+
+"""
+This module provides a way to check to see if flag is available,
+has_flag (built-in to distutils.CCompiler in Python 3.6+), and
+a cpp_flag function, which will compute the highest available
+flag (or if a flag is supported).
+
+LICENSE:
+
+Copyright (c) 2016 Wenzel Jakob <wenzel.jakob@epfl.ch>, All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice, this
+   list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+
+import contextlib
+import os
+import shutil
+import sys
+import tempfile
+import threading
+
+import distutils.errors
+import distutils.command.build_ext
+
+
+WIN = sys.platform.startswith("win")
+PY2 = sys.version_info[0] < 3
+
+
+# It is recommended to use PEP 518 builds if using this module. However, this
+# file explicitly supports being copied into a user's project directory
+# standalone, and pulling pybind11 with the deprecated setup_requires feature.
+# If you copy the file, remember to add it to your MANIFEST.in, and add the current
+# directory into your path if it sits beside your setup.py.
+
+
+# Just in case someone clever tries to multithread
+tmp_chdir_lock = threading.Lock()
+
+
+@contextlib.contextmanager
+def tmp_chdir():
+    "Prepare and enter a temporary directory, cleanup when done"
+
+    # Threadsafe
+    with tmp_chdir_lock:
+        olddir = os.getcwd()
+        try:
+            tmpdir = tempfile.mkdtemp()
+            os.chdir(tmpdir)
+            yield tmpdir
+        finally:
+            os.chdir(olddir)
+            shutil.rmtree(tmpdir)
+
+
+class build_ext(distutils.command.build_ext.build_ext):  # noqa: N801
```

Comment:
I'm actually not a fan of this. It's useful for projects that only build pybind11 extensions, but is hostile to other projects, especially if they are already overriding build_ext.

I think a better approach would be to create a subclass of `Extension` (maybe `Pybind11Extension`?) that adds in the extra flags need by pybind11 to only those extensions, and not globally. See setuptools_dso for an example.

Reply:
I might be able to improve this a bit, but I think there's a reason _everybody_ I've looked at does this, including `setuptools_dso` - The "Extension" class is little more than a named dictionary, and has no access to the compiler for checking to see if flags are supported, or for caching the supported flags for all extensions. `setuptools_dso` hides this a bit by overriding `setup()`, which is even worse. But I can make a PybindExtension that doesn't affect normal extensions.

Reply:
(happy to be proved wrong)

Reply:
A simple workaround would be 

```python
class Pybind11Extension(Extension): pass
```

Then only apply the flags to things that are `isinstance(ext, Pybind11Extension)`.

I'm not very familiar about the internal machinations of setuptools (honestly I wish no one ever had to be), so if this wouldn't work, consider the issue dropped since I'm not qualified enough to make further suggestions.

Reply:
Okay, I've come up with something fairly nice and Extension focused. `build_ext` is only required if you want the automatic C++ flags (since you cannot access the compiler outside of `build_ext`), but the rest are deducible from the platform and Python version. Most users can probably just specify the version they are targeting and forget `build_ext` - I just happen to have a use case in Part 2 for the auto search, and that was what was done in the original example, as well. You can modify the flags (even remove them) after making a Pybind11Extension, too. :)

Reply:
Python 2 + distutils is like pulling teeth. It uses old-style classes everywhere...

Reply:
Thank you! We are hoping to move to pybind11 in [cocotb](https://github.com/cocotb/cocotb) one day. There is already a fair amount of hackery going on with the distutils/setuptools-based build system and we will also be adding setuptools_dso soon as well. I need a solution that is compatible with all of that. This seems like it will be useful now, I hope it will be to other projects as well.